app/controllers/user_controller.rb

```diff
@@ -0,0 +1,112 @@
+require "sinatra/base"
+require "sinatra/activerecord"
+require "bcrypt"
+require_relative "../models/user"
+require_relative "./application_controller"
+require_relative "../helpers/session_helper"
+
+class UserController < Sinatra::Base
+  include SessionHelper
+  enable :sessions
+  register Sinatra::ActiveRecordExtension
+  configure do
+    set :views, 'app/views'
+  end 
+
+  get "/sign_in" do 
+    if logged_in?
+      erb(:redirect_to_user_page)
+    else
+      erb(:sign_in)
+    end 
+  end 
+    
+  post '/sign_in' do
+    user = User.find_by(username: params[:username])
+    if user&.authenticate(params[:password])
+      session[:user_id] = user.id
+      erb(:redirect_to_user_page)
+    else
+      params[:username].empty? || params[:password].empty? 
+      erb(:sign_in_blank)
+    end
+  end
+
```

Comment:
Methods are really concise and I haven't seen any repetition which is really great!